pkg/provisioners/message.go

```diff
@@ -62,3 +72,58 @@ func headerSet(headers []string) map[string]bool {
 	}
 	return set
 }
+
+// History returns the list of hosts where the message has been into
+func (m *Message) History() []string {
+	if m.Headers == nil {
+		return nil
+	}
+	if h, ok := m.Headers[MessageHistoryHeader]; ok {
+		return decodeMessageHistory(h)
+	}
+	return nil
+}
+
+// AppendToHistory append a new host at the end of the list of hosts of the message history
+func (m *Message) AppendToHistory(host string) {
+	host = cleanupMessageHistoryItem(host)
+	if host == "" {
+		return
+	}
+	history := m.History()
+	if history == nil {
+		history = make([]string, 0)
+	}
+	newHistory := append(history, host)
+	m.SetHistory(newHistory)
+}
+
+// SetHistory sets the message history to the given value
+func (m *Message) SetHistory(history []string) {
+	historyStr := encodeMessageHistory(history)
+	if m.Headers == nil {
+		m.Headers = make(map[string]string)
+	}
+	m.Headers[MessageHistoryHeader] = historyStr
+}
+
+func cleanupMessageHistoryItem(host string) string {
+	return strings.Trim(host, " ")
+}
+
+func encodeMessageHistory(history []string) string {
+	return strings.Join(history, MessageHistorySeparator)
+}
+
+func decodeMessageHistory(historyStr string) []string {
+	readHistory := historySplitter.Split(historyStr, -1)
+	// Filter and cleanup in-place
+	history := readHistory[:0]
+	for _, item := range readHistory {
+		cleanItem := cleanupMessageHistoryItem(item)
```

Comment:
Do you need to do this with the historySplitter regexp (other than trimming the spaces at the beginning and end of the string)?

The `\s*` was intended to match spaces before/after `;`.